source/slang/slang-ir-metal-legalize.cpp

```diff
@@ -306,15 +309,110 @@ namespace Slang
+    {
+        auto func = entryPoint.entryPointFunc;
+
+        if (entryPoint.entryPointDecor->getProfile().getStage() != Stage::Mesh)
+        {
+            return;
+        }
+
+        IRBuilder builder{ entryPoint.entryPointFunc->getModule() };
+        for (auto param : func->getParams())
+        {
+            if(param->findDecorationImpl(kIROp_HLSLMeshPayloadDecoration))
+            {
+                IRVarLayout::Builder varLayoutBuilder(&builder, IRTypeLayout::Builder{&builder}.build());
+
+                varLayoutBuilder.findOrAddResourceInfo(LayoutResourceKind::MetalPayload);
+                auto paramVarLayout = varLayoutBuilder.build();
+                builder.addLayoutDecoration(param, paramVarLayout);
+            }
+        }
+
+    }
+
+    void legalizeDispatchMeshPayloadForMetal(EntryPointInfo entryPoint)
+    {
+        if (entryPoint.entryPointDecor->getProfile().getStage() != Stage::Amplification)
+        {
+            return;
+        }
+        // Find out DispatchMesh function
+        IRGlobalValueWithCode* dispatchMeshFunc = nullptr;
+        for (const auto globalInst : entryPoint.entryPointFunc->getModule()->getGlobalInsts())
+        {
+            if (const auto func = as<IRGlobalValueWithCode>(globalInst))
+            {
+                if (const auto dec = func->findDecoration<IRKnownBuiltinDecoration>())
+                {
+                    if (dec->getName() == "DispatchMesh")
+                    {
+                        SLANG_ASSERT(!dispatchMeshFunc && "Multiple DispatchMesh functions found");
+                        dispatchMeshFunc = func;
+                    }
+                }
+            }
+        }
+
+        if (!dispatchMeshFunc)
+            return;
+
+        IRBuilder builder{ entryPoint.entryPointFunc->getModule() };
+
+        // We'll rewrite the call to use mesh_grid_properties.set_threadgroups_per_grid
+        traverseUses(dispatchMeshFunc, [&](const IRUse* use) {
+            if (const auto call = as<IRCall>(use->getUser()))
+            {
+                SLANG_ASSERT(call->getArgCount() == 4);
+                const auto payload = call->getArg(3);
+
+                const auto payloadPtrType = composeGetters<IRPtrType>(
+                    payload,
+                    &IRInst::getDataType
+                );
+                SLANG_ASSERT(payloadPtrType);
+                const auto payloadType = payloadPtrType->getValueType();
+                SLANG_ASSERT(payloadType);
+
+                builder.setInsertBefore(entryPoint.entryPointFunc->getFirstBlock()->getFirstOrdinaryInst());
+                const auto annotatedPayloadType =
+                    builder.getPtrType(
+                        kIROp_RefType,
+                        payloadPtrType->getValueType(),
+                        AddressSpace::MetalObjectData
+                    );
+                auto packedParam = builder.emitParam(annotatedPayloadType);
```

Comment:
If there are more than one DispatchMesh call, this will insert duplicate parameters into the entry point. We should probably make sure we insert at most once for the referencing entry point?

Reply:
Also there could be more than one entry points in a single module. I feel like instead of inserting into the entry point, we should try to find the parent function of the call site, assert that it is a object shader entry point, and insert a param into the EntryPoint if it is t already there.

Reply:
`auto parentFunc = getParentFunc(call)` should give you the user func, you can then `builder.setInsertInto(parentFunc->getFirstBlock()->getFirstOrdinaryInst())` for the param.

Reply:
So we still assume that the DispatchMesh call is done directly from inside the entry point, but we make sure that the payload parameter is only generated once per entrypoint?

Reply:
Yes, let's do that for now. we can add a diagnostic if oarent func isnt an entrypoint.